File: src/main/java/cn/labzen/web/response/format/StandardResultResponseFormatter.java
package cn.labzen.web.response.format;

import cn.labzen.web.paging.Pagination;
import cn.labzen.web.response.bean.Meta;
import cn.labzen.web.response.bean.Response;
import cn.labzen.web.response.bean.Result;
import com.google.common.primitives.Longs;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import java.util.Optional;

import static cn.labzen.web.defination.Constants.REST_REQUEST_TIME;
import static cn.labzen.web.defination.Constants.REST_REQUEST_TIME_MILLIS;

/**
 * 处理 {@link Result} 中的返回值格式化，通用的格式化器，在 {@link CompositeResponseFormatter} 中的执行顺序最晚，如果其他的格式化器不处理返回值，则在这里格式化，算是最后托底的
 */
public class StandardResultResponseFormatter implements ResponseFormatter {

  @Override
  public boolean support(Class<?> clazz, HttpServletRequest request) {
    return Result.class.isAssignableFrom(clazz);
  }

  @Override
  public Object format(Object result, HttpServletRequest request, HttpServletResponse response) {
    String requestTime = request.getAttribute(REST_REQUEST_TIME).toString();
    String requestMillsStr = request.getAttribute(REST_REQUEST_TIME_MILLIS).toString();
    long requestMills = Optional.ofNullable(Longs.tryParse(requestMillsStr)).orElse(0L);
    long executionTime = System.currentTimeMillis() - requestMills;

    Result data = (Result) result;

    int code = data.code();
    String message = Optional.ofNullable(data.message()).orElse("success");
    Object value = data.value();

    if (value instanceof Pagination<?> pagination) {
      Meta meta = new Meta(requestTime, executionTime, pagination.pageable() ? pagination.copyWithoutRecords() : null, null, null);
      return new Response(code, message, meta, pagination.records());
    } else {
      Meta meta = new Meta(requestTime, executionTime, null, null, null);
      return new Response(code, message, meta, value);
    }
  }
}

File: src/main/kotlin/cn/labzen/web/response/format/StandardResultResponseFormatter.kt
package cn.labzen.web.response.format

import cn.labzen.web.REST_REQUEST_TIME
import cn.labzen.web.REST_REQUEST_TIME_MILLIS
import cn.labzen.web.paging.Pagination
import cn.labzen.web.response.bean.Meta
import cn.labzen.web.response.bean.Response
import cn.labzen.web.response.bean.Result
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 处理 [Result] 中的返回值格式化，通用的格式化器，在 [CompositeResponseFormatter] 中的执行顺序最晚，如果其他的格式化器不处理返回值，则在这里格式化，算是最后托底的
 */
class StandardResultResponseFormatter : ResponseFormatter {

  override fun support(clazz: Class<*>, request: HttpServletRequest): Boolean =
    clazz == Result::class.java

  override fun format(result: Any?, request: HttpServletRequest, response: HttpServletResponse): Any {
    val requestTime = request.getAttribute(REST_REQUEST_TIME).toString()
    val requestMillis = request.getAttribute(REST_REQUEST_TIME_MILLIS).toString().toLong()
    val executionTime = System.currentTimeMillis() - requestMillis

    val data = result as Result

    val code = data.code
    val message = data.message ?: "success"
    val value = data.value

    return if (value is Pagination<*>) {
      val records = value.records
      value.records = null
      val meta = Meta(requestTime, executionTime, value)
      Response(code, message, meta, records)
    } else {
      val meta = Meta(requestTime, executionTime)
      Response(code, message, meta, value)
    }
  }
}